File: 백준/Gold/17281. ⚾/⚾.py
from sys import stdin
input = stdin.readline

n = int(input())
a = [list(map(int, input().split())) for _ in range(n)]
select, check = [0 for _ in range(9)], [0 for _ in range(9)]
select[3], check[3] = 0, 1


def game():
    global ans
    cur, score = 0, 0
    for inning in a:
        out, b1, b2, b3 = 0, 0, 0, 0
        while out <= 2:
            r = inning[select[cur]]
            if r == 0:
                out += 1
            elif r == 1:
                score += b3
                b1, b2, b3 = 1, b1, b2
            elif r == 2:
                score += b2 + b3
                b1, b2, b3 = 0, 1, b1
            elif r == 3:
                score += b1 + b2 + b3
                b1, b2, b3 = 0, 0, 1
            else:
                score += b1 + b2 + b3 + 1
                b1, b2, b3 = 0, 0, 0
            cur += 1
            cur %= 9
    ans = max(ans, score)
    return


def solve(cnt):
    if cnt == 9:
        game()
    for i in range(9):
        if check[i]:
            continue
        check[i] = 1
        select[i] = cnt
        solve(cnt + 1)
        check[i] = 0
        select[i] = 0


ans = 0
solve(1)
print(ans)

File: 백준/Gold/17281. ⚾/⚾.java
import java.io.*;
import java.util.*;

public class Main {
	static int[][] play;
	static int[] seq;
	static boolean[] v;
	static int N, playerIdx, ans;

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		N = Integer.parseInt(br.readLine());
		play = new int[N][9];
		seq = new int[9];
		v = new boolean[9];
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < 9; j++) {
				play[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		ans = 0;
		perm(0);
		System.out.println(ans);
	}

	static void perm(int cnt) {
		if (cnt == 9) {
			ans = Math.max(ans, game());
			return;
		}
		if (cnt == 3) {
			seq[cnt] = 0;
			perm(cnt + 1);
			return;
		}
		for (int i = 1; i < 9; i++) {
			if (v[i])
				continue;
			v[i] = true;
			seq[cnt] = i;
			perm(cnt + 1);
			v[i] = false;
		}
	}

	static int game() {
		playerIdx = 0;
		int score = 0;
		int turn = 0;
		while (turn != N) {
			int flag = 0;
			int outCnt = 0;
			while (true) {
				int action = play[turn][seq[playerIdx]];
				playerIdx = (playerIdx + 1) % 9;
				if (action == 0) {
					outCnt++;
					if (outCnt == 3) {
						outCnt = 0;
						break;
					}
				} else {
					for (int i = 0; i < action; i++) {
						if (i == 0) {
							flag <<= 1;
							flag += 1;
						} else
							flag <<= 1;
						if ((flag & (1 << 3)) != 0) {
							flag -= (1 << 3);
							score++;
						}
					}
				}
			}
			turn++;
		}
		return score;
	}
}